### PR TITLE
Rapid packet-based entity data changes

### PR DESCRIPTION
**Pull request by tal5**
## Changes

- `EntityHelper#modifyInternalEntityData` now takes in a `MapTag`, to match the new method & share it's internal handling.
- Minor update to entity data ID mapping, display entities have a new one now.
- The new `PacketHelperImpl(1.20)#sendAsyncSafe`'s code was slightly changed to work properly without network interception being enabled.

## Removals

- `EntityHelper#mapInternalEntityDataName` was removed in favor of having that logic/handling in NMS.

## Additions

- `EntityHelper#convertInternalEntityDataValues` - converts a `MapTag` to a list of `DataValue`'s (returned as `List<Object>`), used to parse entity data in the new command, implemented as such a bridge method so that it can be done outside of the async task.
- `PacketHelperImpl#sendEntityDataPacket` - sends an entity data update packet to a list of players, taking in a `List<Object>` (`DataValue`s).
- `fakeinternaldata` command - sends fake entity data updates, optionally animating them with sub-tick precision.
- `EntityHelperImpl(1.20)#getDataItems` - util method to get an entity's internal `id -> DataItem` map.
- `EntityHelperImpl(1.20)#convertToInternalData` - internal util to parse a `MapTag` into entity data values, taking in a `BiConsumer` to use the converted values.
- `PacketHelperImpl(1.20)#sendAsyncSafe` - util to send packets async, as this was done in 2 places now.
- `DenizenNetworkManagerImpl#getConnection(Player)` - util to get a connection from a bukkit player.
- `Internal Entity Data` language doc - a general explanation of internal entity data with relevant links, used in both the command and the mechanism.

## Notes

- Did my best to make this relatively clean, but overall it's still a pretty messy setups doing to passing NMS objects around outside of NMS and all; sadly there isn't really a way to make this nicer other than having all of the logic in NMS (or making some over the top API for ourselves), but let me know if that should be done.
- The new command can technically be used to just send a fake entity data update to a player (which is why I've made `speed` optional), but is that something we want to officially document? if we do we should probably
  - Add a special case in the code to just send the update instead of doing the entire async loop thingy for a single packet. 
  - What about future updates messing it up? most similar commands have packet interception to persist themselves.

### CODE REVIEW

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/BukkitCommandRegistry.java [hunk 1]
```diff
@@ -85,6 +85,7 @@ public static void registerCommands() {
         registerCommand(ResetCommand.class);
         registerCommand(ZapCommand.class);
         // entity
+        registerCommand(SmoothEntityDataCommand.class); // <---------
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/BukkitCommandRegistry.java]: why the comment

File: v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityHelperImpl.java [hunk 2]
```diff
@@ -806,6 +806,27 @@ public int mapInternalEntityDataName(Entity entity, String name) {
         return EntityDataNameMapper.getIdForName(((CraftEntity) entity).getHandle().getClass(), name);
     }
 
+    public static final Field SynchedEntityData_itemsById = ReflectionHelper.getFields(SynchedEntityData.class).get(ReflectionMappingsInfo.SynchedEntityData_itemsById);
+
+    public static Int2ObjectMap<SynchedEntityData.DataItem<Object>> getDataItems(Entity entity) {
```
Comment (mcmonkey4eva) [inline v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityHelperImpl.java:807]: should update `modifyInternalEntityData` to use this

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/BukkitCommandRegistry.java (hunk 1 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/BukkitCommandRegistry.java]: Will remove later, was a ~~badly made~~ TODO for myself

File: v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityHelperImpl.java (hunk 2 above)
Comment (tal5) [reply v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityHelperImpl.java:807]: Ye, haven't touched the existing `internal_data` mechanism yet as I haven't finished finalizing how the new methods are setup

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FakeInternalDataCommand extends AbstractCommand {
+
+    public FakeInternalDataCommand() {
+        setName("fakeinternaldata");
+        setSyntax("fakeinternaldata [entity:<entity>] [data:<map>|...] [for:<player>|...] (speed:<duration>)");
```
Comment (mergu) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: `for` arg should be optional and default to linked player

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import org.bukkit.entity.Player;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class FakeInternalDataCommand extends AbstractCommand {
+
+    public FakeInternalDataCommand() {
```
Comment (mergu) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:39]: meta missing?

Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:39]: Ye, PR is still a draft - will add it once we have the command itself finalized

File: v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityHelperImpl.java (hunk 2 above)
Comment (tal5) [reply v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityHelperImpl.java:807]: This is cleaned up now

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 4 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:39]: Added meta

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class FakeInternalDataCommand extends AbstractCommand {
+
+    public FakeInternalDataCommand() {
+        setName("fakeinternaldata");
+        setSyntax("fakeinternaldata [entity:<entity>] [data:<map>|...] (for:<player>|...) (speed:<duration>)");
+        setRequiredArguments(3, 4);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: I think `3` is wrong here?

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    // 'speed:' is the amount of time between each frame getting sent, supporting sub-tick delays.
+    //
+    // @Usage
+    // Animates an item display entity's item for the linked player, and slowly scales it up.
+    // - fakeinternaldata entity:<[item_display]> data:[item=iron_ingot;scale=0.6,0.6,0.6]|[item=gold_ingot;scale=0.8,0.8,0.8]|[item=netherite_ingot;scale=1,1,1] speed:0.5s
+    //
+    // @Usage
+    // Animates a rainbow glow on a display entity for all online players.
+    // - define color <color[red]>
+    // - repeat 361 from:0 as:hue:
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: A: hue is 0-255, not 0-360?
B: even if it wasn't a byte, the usage of `361` would still be 1 too many here (in a circle, 0 degrees and 360 degrees are the same value)

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    //   - define frames:->:<map[glow_color=<[color].with_hue[<[hue]>].argb_integer>]>
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: constructor tag is redundant

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        List<List<Object>> frames = new ArrayList<>(data.size());
+        for (MapTag frame : data) {
+            frames.add(NMSHandler.entityHelper.convertInternalEntityDataValues(entity, frame));
+        }
+        long ms = speed.getMillis();
+        DenizenCore.runAsync(() -> {
+            try {
+                for (List<Object> frame : frames) {
+                    NMSHandler.packetHelper.sendEntityDataPacket(sendTo, entity, frame);
+                    Thread.sleep(ms);
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Should either note in meta that the delay is inaccurate for long runs, or switch this sleep to a more accurate delay manager.

Also considering the possibility of long runs, should also add a check to kill the loop if the player(s) are gone

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 5 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Forgot to update that, fixed

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 6 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Should be corrected now

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 7 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Should be corrected now

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 8 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Both should be fixed

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        setName("fakeinternaldata");
+        setSyntax("fakeinternaldata [entity:<entity>] [data:<map>|...] (for:<player>|...) (speed:<duration>)");
+        setRequiredArguments(2, 4);
+        autoCompile();
+    }
+
+    // <--[command]
+    // @Name FakeInternalData
+    // @Syntax fakeinternaldata [entity:<entity>] [data:<map>|...] (for:<player>|...) (speed:<duration>)
+    // @Required 3
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: required min mismatch between meta and setSyntax

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    //
+    // The input to 'data:' is a list of <@link object MapTag>s, with each map being a frame to send, with each map being formatted like <@link mechanism EntityTag.internal_data>'s input.
+    //
+    // Optionally specify a list of players to fake the data for, defaults to the linked player.
+    //
+    // 'speed:' is the amount of time between each frame getting sent, supporting sub-tick delays.
+    //
+    // @Usage
+    // Animates an item display entity's item for the linked player, and slowly scales it up.
+    // - fakeinternaldata entity:<[item_display]> data:[item=iron_ingot;scale=0.6,0.6,0.6]|[item=gold_ingot;scale=0.8,0.8,0.8]|[item=netherite_ingot;scale=1,1,1] speed:0.5s
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:70]: The meta should indicate clearly what would happen if you do eg `data:[item=iron_ingot]|[scale=2,2,2]` (ie use different keys between frames)

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+    // @Required 3
+    // @Maximum 4
+    // @Short Sends fake entity data updates, optionally animating them with sub-tick precision.
+    // @Group entity
+    //
+    // @Description
+    // Sends fake internal entity data updates, optionally sending multiple over time.
+    // This supports sub-tick precision, allowing smooth/high FPS animations.
+    //
+    // The input to 'data:' is a list of <@link object MapTag>s, with each map being a frame to send, with each map being formatted like <@link mechanism EntityTag.internal_data>'s input.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: A command should probably be considered "more primary" than a mech (ie put the details in command and have the mech link to it) - or add a separate `language` defining the data that both link to

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+public class FakeInternalDataCommand extends AbstractCommand {
+
+    public FakeInternalDataCommand() {
+        setName("fakeinternaldata");
+        setSyntax("fakeinternaldata [entity:<entity>] [data:<map>|...] (for:<player>|...) (speed:<duration>)");
+        setRequiredArguments(2, 4);
+        autoCompile();
+    }
+
+    // <--[command]
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:46]: need `@Tags` - can use `None` if there's no applicable tags (if so - why is there no tag to read this data?)

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 9 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Fixed

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 10 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:70]: Tried explaining that & added an example, lmk if that looks good

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 12 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:46]: Added

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+    //
+    // @Tags
+    // None
+    //
+    // @Usage
+    // Animates an item display entity's item for the linked player, and slowly scales it up.
+    // - fakeinternaldata entity:<[item_display]> data:[item=iron_ingot;scale=0.6,0.6,0.6]|[item=gold_ingot;scale=0.8,0.8,0.8]|[item=netherite_ingot;scale=1,1,1] speed:0.5s
+    //
+    // @Usage
+    // Changes an item display's item, then it's scale a second later, then it's item again another second later.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: `its`

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 11 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Pushed an initial attempt at a language doc for this, let me know if it looks good

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java (hunk 13 above)
Comment (tal5) [reply plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java]: Fixed

File: plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+public class FakeInternalDataCommand extends AbstractCommand {
+
+    // <--[language]
+    // @name Internal Entity Data
+    // @group Minecraft Logic
+    // @description
+    // Each entity in Minecraft has a set of data values that get sent to the client, with each data value being a number id -> value pair.
+    // Denizen allows direct control over that data, as it can be useful for things like setting values that would usually be blocked.
+    // Because this is such a direct control that's meant to impose less restrictions, there's no limitations/verification on the values being set other than basic type checking.
+    // For all possible internal entity data values and their respective ids, see <@link url https://github.com/DenizenScript/Denizen/blob/dev/v1_20/src/main/java/com/denizenscript/denizen/nms/v1_20/helpers/EntityDataNameMapper.java#L50>.
```
Comment (mcmonkey4eva) [inline plugin/src/main/java/com/denizenscript/denizen/scripts/commands/entity/FakeInternalDataCommand.java:32]: This doc should note somewhere that these values are liable to change between minecraft versions

**Pull request merged**